hw/ip/prim/util/secded_gen.py

```diff
@@ -177,28 +214,72 @@ def main():
     fanin_ideal = ideal_fanin(k, m)
     log.info("Ideal Fan-In value: %d" % fanin_ideal)
 
+    # Each entry represents a row in below parity matrix
+    # Entry is tuple and the value inside is the position of ones
+    # e.g. (0,1,2) in m:=7
+    # row -> [1 1 1 0 0 0 0]
     codes = []
 
+    ## Find code matrix =======================================================
+    # This is main part to find the parity matrix.
+    # For example, find SECDED for 4bit message is to find 3x4 matrix as below
+    # | 1 0 0 0 x x x |
+    # | 0 1 0 0 x x x |
+    # | 0 0 1 0 x x x |
+    # | 0 0 0 1 x x x |
+    # Then message _k_ X matrix_code ==> original message with parity
+    #
+    # Make a row to have even number of 1 including the I matrix.
+    # This helps to calculate the syndrom at the decoding stage.
+    # To reduce the max fan-in, Starting with smallest number 3.
+    # the number means the number of one in a row.
+    # Small number of ones means smaller fan-in overall.
+
     for step in range(3, m + 1, 2):
         # starting from 3 as I matrix represents data
         # Increased by 2 as number of 1 should be even in a row (odd excluding I)
 
-        # get the list of combinations
+        # get the list of combinations [0, .., m-1] with `step`
+        # e.g. step := 3 ==> [(0,1,2), (0,1,3), ... ]
         candidate = list(itertools.combinations(range(m), step))
 
-        # Let's shuffle
-        random.shuffle(candidate)
-
         if len(candidate) <= required_row:
             # we need more round use all of them
             codes.extend(candidate)
             required_row -= len(candidate)
         else:
-            # we can completed in this round
-            # but search lowest fan-in codes
-            # at this time, just pick lowest
-            codes.extend(candidate[0:required_row])
-            required_row = 0
+            ## Find optimized fan-in ==========================================
+
+            # Calculate each row fan-in with current
+            fanins = calc_fanin(m, codes)
+            while required_row != 0:
```

Comment:
am I right that this tries to find an optimal subset of candidate rows?

looks like an [assignment problem](https://en.wikipedia.org/wiki/Assignment_problem) that could be solved using a variant of the Hungarian method (maybe there is even an optimization library for this).

how long does this take to converge? If it's just a matter of seconds I would leave it as-is, since implementing the above may not be straightforward...